laws/src/main/scala/scalaz/tests/TypeclassLaws.scala

```diff
@@ -0,0 +1,234 @@
+    )
+  }
+}
+
+// the other "composition" laws for all kinds of functors are guaranteed by parametricity
+// in addition to the "identity" law.
+object FunctorLaws {
+  @inline
+  def identityToIdentity[F[_], A, T](in: F[A])(assert: (F[A], F[A]) => T)(implicit F: Functor[F]): T =
+    assert(in, F.map(in)(identity))
+}
+
+object ContravariantLaws {
+  @inline
+  def identityToIdentity[F[_], A, T](in: F[A])(assert: (F[A], F[A]) => T)(implicit F: Contravariant[F]): T =
+    assert(in, F.contramap(in)(identity))
+}
+
+object InvariantFunctorLaws {
+  @inline
+  def identityToIdentity[F[_], A, T](in: F[A])(assert: (F[A], F[A]) => T)(implicit F: InvariantFunctor[F]): T =
+    assert(in, F.imap(in)(identity)(identity))
+}
+
+object PhantomLaws {
+  @inline
+  def identityToIdentity[F[_], A, T](in: F[A])(assert: (F[A], F[A]) => T)(implicit F: Phantom[F]): T =
+    assert(in, F.pmap(in))
+}
+
+object BifunctorLaws {
+  @inline
+  def bimapIdentityToIdentity[F[_, _], A, B, T](
+    in: F[A, B]
+  )(assert: (F[A, B], F[A, B]) => T)(implicit F: Bifunctor[F]): T =
+    assert(in, F.bimap(in)(identity, identity))
+}
+
+object BindLaws {
+  @inline
+  def bindAssoc[F[_], A, B, C, T](in: F[A])(fst: A => F[B],
+                                            snd: B => F[C])(assert: (F[C], F[C]) => T)(implicit F: Bind[F]): T = {
+    import F.flatMap
+    assert(
+      flatMap(flatMap(in)(fst))(snd),
+      flatMap(in)(b => flatMap(fst(b))(snd))
+    )
+  }
+}
+
+object SemicategoryLaws {
+  @inline
+  def composeAssoc[F[_, _], A, B, C, D, T](fst: F[C, D], snd: F[B, C], thd: F[A, B])(
+    assert: (F[A, D], F[A, D]) => T
+  )(implicit F: Semicategory[F]): T = {
+    import F.compose
+    assert(
+      compose(compose(fst, snd), thd),
+      compose(fst, compose(snd, thd))
+    )
+  }
+}
+
+object CategoryLaws {
+  @inline
+  def composeRightIdentity[F[_, _], A, B, T](in: F[A, B])(assert: (F[A, B], F[A, B]) => T)(implicit F: Category[F]): T =
+    assert(in, F.compose(in, F.id))
+
+  @inline
+  def composeLeftIdentity[F[_, _], A, B, T](in: F[A, B])(assert: (F[A, B], F[A, B]) => T)(implicit F: Category[F]): T =
+    assert(in, F.compose(F.id, in))
+}
+
+object MonadLaws {
+  @inline
+  def bindIdentity[F[_], A, T](in: F[A])(assert: (F[A], F[A]) => T)(implicit F: Monad[F]) =
+    assert(in, F.flatMap(in)(F.pure))
+}
+
+object ApplyLaws {
+  @inline
+  def applyAssoc[F[_], A, B, C, T](in: F[A])(fst: F[A => B],
+                                             snd: F[B => C])(assert: (F[C], F[C]) => T)(implicit F: Apply[F]): T = {
+    import F.{ ap, map }
+    def compose(f: B => C)(g: A => B): A => C = f compose g
+    assert(
+      ap(ap(in)(fst))(snd),
+      ap(in)(ap(fst)(map(snd)(compose)))
+    )
+  }
+}
+
+object ApplicativeLaws {
+  @inline
+  def applyIdentity[F[_], A, B, T](in: A)(f: A => B)(assert: (F[B], F[B]) => T)(implicit F: Applicative[F]) =
+    assert(
+      F.pure(f(in)),
+      F.ap(F.pure(in))(F.pure(f))
+    )
+}
+
+object EqLaws {
```

Comment:
Reflexivity? Transitivity and symmetry follow from the indiscernability + reflexivity.

Reply:
Will do.

Reply:
Done.